rules/DeadCode/Rector/FunctionLike/NarrowTooWideReturnTypeRector.php

```diff
@@ -0,0 +1,253 @@
+            : $this->betterNodeFinder->findReturnsScoped($node);
+        $isAlwaysTerminating = ! $this->silentVoidResolver->hasSilentVoid($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            return null;
+        }
+
+        $returnType = $node->returnType;
+        Assert::isInstanceOfAny($returnType, [UnionType::class, NullableType::class]);
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowUnionReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $newReturnType;
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType && ! $returnType instanceof NullableType) {
+            return true;
+        }
+
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        if (! $classReflection->isClass()) {
+            return true;
+        }
+
+        return ! $classReflection->isFinalByKeyword();
+    }
+
+    /**
+     * @param Return_[] $returnStatements
+     * @return Type[]
+     */
+    private function collectActualReturnTypes(
+        ClassMethod|Function_|Closure|ArrowFunction $node,
+        array $returnStatements,
+        bool $isAlwaysTerminating,
+    ): array {
+        if ($node instanceof ArrowFunction) {
+            return [$this->nodeTypeResolver->getNativeType($node->expr)];
+        }
+
+        $returnTypes = [];
+        foreach ($returnStatements as $returnStatement) {
+            if ($returnStatement->expr === null) {
+                $returnTypes[] = new NullType();
+                continue;
+            }
+
+            $returnTypes[] = $this->nodeTypeResolver->getNativeType($returnStatement->expr);
+        }
+
+        if (! $isAlwaysTerminating) {
+            $returnTypes[] = new NullType();
+        }
+
+        return $returnTypes;
+    }
+
+    /**
+     * @param Type[] $actualReturnTypes
+     */
+    private function narrowUnionReturnType(
+        UnionType|NullableType $returnType,
+        array $actualReturnTypes
+    ): ComplexType|Identifier|Name|null {
+        $types = $returnType instanceof UnionType
+            ? $returnType->types
+            : [$returnType->type, new Identifier('null')];
```

Comment:
You should can actually directly convert it into PHPStan type.

```php
$type = $this->staticTypeMapper->mapPhpParserNodePHPStanType($returnType);
```

it may need some tweak as you do tho, that check its `$type->getTypes()`

Reply:
Done!